src/osyris/core/subdomain.py

```diff
@@ -0,0 +1,159 @@
+                    valid_comp_amr = (valid_comp_x_amr & valid_comp_y_amr
+                                      & valid_comp_z_amr).values
+                    # find sink & particles indices
+                    if self._sink_particles:
+                        valid_comp_x_sinks = (centered_sinks_pos.x >= -dx * .5) & (
+                            centered_sinks_pos.x <= dx * .5)
+                        valid_comp_y_sinks = (centered_sinks_pos.y >= -dy * .5) & (
+                            centered_sinks_pos.y <= dy * .5)
+                        valid_comp_z_sinks = (centered_sinks_pos.z >= -dz * .5) & (
+                            centered_sinks_pos.z <= dz * .5)
+                        valid_comp_sinks = (valid_comp_x_sinks & valid_comp_y_sinks
+                                            & valid_comp_z_sinks).values
+                        if not np.any(valid_comp_sinks):
+                            self._sink_particles = False
+                    if self._particles:
+                        valid_comp_x_particles = (centered_particles_pos.x >=
+                                                  -dx * .5) & (centered_particles_pos.x
+                                                               <= dx * .5)
+                        valid_comp_y_particles = (centered_particles_pos.y >=
+                                                  -dy * .5) & (centered_particles_pos.y
+                                                               <= dy * .5)
+                        valid_comp_z_particles = (centered_particles_pos.z >=
+                                                  -dz * .5) & (centered_particles_pos.z
+                                                               <= dz * .5)
+                        valid_comp_particles = (valid_comp_x_particles
+                                                & valid_comp_y_particles
+                                                & valid_comp_z_particles).values
+                        if not np.any(valid_comp_particles):
+                            self._particles = False
+                else:
+                    raise ValueError("Please specify a valid dx, dy and dz")
+            else:
+                raise ValueError(
+                    "Unrecognized string '{}', valid criterions are 'spherical' and 'cubic'"
+                    .format(selection.lower()))
+        elif isinstance(selection, np.array):
+            raise NotImplementedError(
+                "Numpy array indexing for subdomain extraction coming soon...")
+        else:
+            raise ValueError(
+                "Please specify a valid selection criterion (either 'spherical' or 'cubic')"
+            )
+
+        if not np.any(valid_comp_amr):
+            raise ValueError("Empty domain")
+
+        return valid_comp_amr, valid_comp_sinks, valid_comp_particles
+
+    def _extract_data(self, valid_comp_amr, valid_comp_sinks, valid_comp_particles):
```

Comment:
Not sure why you need the two steps of first finding indices, and then extracting. Can't you do it all in one go, as suggested above?

Reply:
Yeah I'll do that.